### PR TITLE
add event source on Lambda examples Gen 2

### PR DESCRIPTION
**Pull request by ykethan**
#### Description of changes:

provides an example on using the event source as kinesis stream and DDB stream on a Lambda function

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] Swift
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://docs.amplify.aws/)` 
            HTML: `<a href="https://docs.amplify.aws/">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+```
+
+Second, create a new directory and a resource file, `amplify/functions/kinesis-function/resource.ts`. Then, define the Function with `defineFunction`:
+
+```ts title="amplify/functions/kinesis-function/resource.ts"
+import { defineFunction } from "@aws-amplify/backend";
+
+export const myKinesisFunction = defineFunction({
+  name: "kinesis-function",
+  entry: "./handler.ts",
```
Comment (josefaidt) [inline src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx:49]:
```suggestion
```

File: src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  return { batchItemFailures: [] };
+};
+
+async function getRecordDataAsync(
+  payload: KinesisStreamRecordPayload
+): Promise<string> {
+  var data = Buffer.from(payload.data, "base64").toString("utf-8");
+  await Promise.resolve(1); //Placeholder for an async process
+  return data;
+}
```
Comment (josefaidt) [inline src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx:101]:
```suggestion
import type {
  KinesisStreamBatchResponse,
  KinesisStreamHandler,
  KinesisStreamRecordPayload,
} from "aws-lambda";
import { Buffer } from "node:buffer";
import { Logger } from "@aws-lambda-powertools/logger";

const logger = new Logger({
  logLevel: "INFO",
  serviceName: "kinesis-stream-handler",
});

export const handler: KinesisStreamHandler = async (
  event,
  context
): Promise<KinesisStreamBatchResponse> => {
  for (const record of event.Records) {
    try {
      logger.info(`Processed Kinesis Event - EventID: ${record.eventID}`);
      const recordData = await getRecordDataAsync(record.kinesis);
      logger.info(`Record Data: ${recordData}`);
    } catch (err) {
      logger.error(`An error occurred ${err}`);
      /**
			 * When processing stream data, if any item fails, returning the failed item's position immediately
			 * prompts Lambda to retry from this item forward, ensuring continuous processing without skipping data.
			 */
      return {
        batchItemFailures: [{ itemIdentifier: record.kinesis.sequenceNumber }],
      };
    }
  }
  logger.info(`Successfully processed ${event.Records.length} records.`);
  return { batchItemFailures: [] };
};

async function getRecordDataAsync(
  payload: KinesisStreamRecordPayload
): Promise<string> {
  const data = Buffer.from(payload.data, "base64").toString("utf-8");
  await Promise.resolve(1); // Placeholder for an async process
  return data;
}
```

File: src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import { myKinesisFunction } from "./functions/kinesis-function/resource";
+import { KinesisEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
+import { StartingPosition } from "aws-cdk-lib/aws-lambda";
+
+const backend = defineBackend({
+  auth,
+  data,
+  myKinesisFunction,
+  //other resources
+});
```
Comment (josefaidt) [inline src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx:119]:
```suggestion
import { defineBackend } from "@aws-amplify/backend";
import { KinesisEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
import { StartingPosition } from "aws-cdk-lib/aws-lambda";
import { auth } from "./auth/resource";
import { data } from "./data/resource";
import { myKinesisFunction } from "./functions/kinesis-function/resource";

const backend = defineBackend({
  auth,
  data,
  myKinesisFunction,
});
```

File: src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+export const myDynamoDBFunction = defineFunction({
+  name: "dynamoDB-function",
+});
+```
+
+Third, create the corresponding handler file, `amplify/functions/dynamoDB-function/handler.ts`, file with the following contents:
+
+```ts title="amplify/functions/dynamoDB-function/resource.ts"
+import { Logger } from "@aws-lambda-powertools/logger";
+import { DynamoDBStreamHandler } from "aws-lambda";
```
Comment (josefaidt) [inline src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx:160]:
```suggestion
import type { DynamoDBStreamHandler } from "aws-lambda";
import { Logger } from "@aws-lambda-powertools/logger";
```

explicit type import nit

File: src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+const logger = new Logger({
+  logLevel: "INFO",
+  serviceName: "dynamodb-stream-handler",
+});
+
+export const handler: DynamoDBStreamHandler = async (event) => {
+  const batchItemFailures: { itemIdentifier: string }[] = [];
```
Comment (josefaidt) [inline src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx:168]: is this needed?

File: src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx [hunk 6]
```diff
@@ -0,0 +1,202 @@
+import { getCustomStaticPath } from '@/utils/getCustomStaticPath';
+
+export const meta = {
+  title: 'Lambda event source',
```
Comment (josefaidt) [inline src/pages/[platform]/build-a-backend/functions/examples/event-source/index.mdx:4]: can we change this title to be a bit more descriptive? maybe just "Event handlers"?

**Review by josefaidt — APPROVED** [review]

**Review by timngyn — APPROVED** [review]

**Pull request merged**